// @storybook/ui/dist/components/sidebar/useExpanded.js
"use strict";

require("core-js/modules/es.symbol");

require("core-js/modules/es.symbol.description");

require("core-js/modules/es.symbol.iterator");

require("core-js/modules/es.array.concat");

require("core-js/modules/es.array.from");

require("core-js/modules/es.array.includes");

require("core-js/modules/es.array.is-array");

require("core-js/modules/es.array.iterator");

require("core-js/modules/es.array.reduce");

require("core-js/modules/es.array.slice");

require("core-js/modules/es.date.to-string");

require("core-js/modules/es.function.name");

require("core-js/modules/es.object.assign");

require("core-js/modules/es.object.define-property");

require("core-js/modules/es.object.to-string");

require("core-js/modules/es.regexp.to-string");

require("core-js/modules/es.string.iterator");

require("core-js/modules/web.dom-collections.iterator");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useExpanded = void 0;

var _global = require("global");

var _throttle = _interopRequireDefault(require("lodash/throttle"));

var _react = require("react");

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

var initializeExpanded = function initializeExpanded(_ref) {
  var data = _ref.data,
      highlightedItemId = _ref.highlightedItemId,
      rootIds = _ref.rootIds;
  var highlightedAncestors = highlightedItemId ? (0, _utils.getAncestorIds)(data, highlightedItemId) : [];
  return [].concat(_toConsumableArray(rootIds), _toConsumableArray(highlightedAncestors)).reduce(function (acc, id) {
    return Object.assign(acc, _defineProperty({}, id, true));
  }, {});
};

var useExpanded = function useExpanded(_ref2) {
  var containerRef = _ref2.containerRef,
      isBrowsing = _ref2.isBrowsing,
      refId = _ref2.refId,
      data = _ref2.data,
      rootIds = _ref2.rootIds,
      highlightedItemId = _ref2.highlightedItemId,
      setHighlightedItemId = _ref2.setHighlightedItemId,
      selectedStoryId = _ref2.selectedStoryId,
      onSelectStoryId = _ref2.onSelectStoryId;

  // Track the set of currently expanded nodes within this tree.
  // Root nodes are expanded by default (and cannot be collapsed).
  var _useReducer = (0, _react.useReducer)(function (state, _ref3) {
    var ids = _ref3.ids,
        value = _ref3.value;
    return ids.reduce(function (acc, id) {
      return Object.assign(acc, _defineProperty({}, id, value));
    }, Object.assign({}, state));
  }, {
    data: data,
    highlightedItemId: highlightedItemId,
    rootIds: rootIds
  }, initializeExpanded),
      _useReducer2 = _slicedToArray(_useReducer, 2),
      expanded = _useReducer2[0],
      setExpanded = _useReducer2[1];

  var getElementByDataItemId = (0, _react.useCallback)(function (id) {
    return containerRef.current && containerRef.current.querySelector("[data-item-id=\"".concat(id, "\"]"));
  }, [containerRef]);
  var highlightElement = (0, _react.useCallback)(function (element) {
    setHighlightedItemId(element.getAttribute('data-item-id'));
    (0, _utils.scrollIntoView)(element);
  }, [setHighlightedItemId]); // Expand the whole ancestry of the currently selected story whenever it changes.

  (0, _react.useEffect)(function () {
    setExpanded({
      ids: (0, _utils.getAncestorIds)(data, selectedStoryId),
      value: true
    });
  }, [data, selectedStoryId]); // Expand, collapse or select nodes in the tree using keyboard shortcuts.

  (0, _react.useEffect)(function () {
    var menuElement = _global.document.getElementById('storybook-explorer-menu');

    var navigateTree = (0, _throttle["default"])(function (event) {
      if (!isBrowsing || !event.key || !containerRef.current || !highlightedItemId) return;
      if (event.shiftKey || event.metaKey || event.ctrlKey || event.altKey) return;
      if (!['Enter', ' ', 'ArrowLeft', 'ArrowRight'].includes(event.key)) return;
      var highlightedElement = getElementByDataItemId(highlightedItemId);
      if (!highlightedElement || highlightedElement.getAttribute('data-ref-id') !== refId) return;
      var target = event.target;
      if (!(0, _utils.isAncestor)(menuElement, target) && !(0, _utils.isAncestor)(target, menuElement)) return;

      if (target.hasAttribute('data-action')) {
        if (['Enter', ' '].includes(event.key)) return;
        target.blur();
      }

      event.preventDefault();
      var type = highlightedElement.getAttribute('data-nodetype');

      if (['Enter', ' '].includes(event.key) && ['component', 'story', 'document'].includes(type)) {
        onSelectStoryId(highlightedItemId);
      }

      var isExpanded = highlightedElement.getAttribute('aria-expanded');

      if (event.key === 'ArrowLeft') {
        if (isExpanded === 'true') {
          // The highlighted node is expanded, so we collapse it.
          setExpanded({
            ids: [highlightedItemId],
            value: false
          });
          return;
        }

        var parentId = highlightedElement.getAttribute('data-parent-id');
        var parentElement = parentId && getElementByDataItemId(parentId);

        if (parentElement && parentElement.getAttribute('data-highlightable') === 'true') {
          // The highlighted node isn't expanded, so we move the highlight to its parent instead.
          highlightElement(parentElement);
          return;
        } // The parent can't be highlighted, which means it must be a root.
        // The highlighted node is already collapsed, so we collapse its descendants.


        setExpanded({
          ids: (0, _utils.getDescendantIds)(data, highlightedItemId, true),
          value: false
        });
        return;
      }

      if (event.key === 'ArrowRight') {
        if (isExpanded === 'false') {
          setExpanded({
            ids: [highlightedItemId],
            value: true
          });
        } else if (isExpanded === 'true') {
          setExpanded({
            ids: (0, _utils.getDescendantIds)(data, highlightedItemId, true),
            value: true
          });
        }
      }
    }, 16);

    _global.document.addEventListener('keydown', navigateTree);

    return function () {
      return _global.document.removeEventListener('keydown', navigateTree);
    };
  }, [containerRef, isBrowsing, refId, data, highlightedItemId, setHighlightedItemId, onSelectStoryId]);
  var updateExpanded = (0, _react.useCallback)(function (_ref4) {
    var ids = _ref4.ids,
        value = _ref4.value;
    setExpanded({
      ids: ids,
      value: value
    });
    if (ids.length === 1) setHighlightedItemId(ids[0]);
  }, [setHighlightedItemId]);
  return [expanded, updateExpanded];
};

exports.useExpanded = useExpanded;
